cpp/src/binaryop/binaryop.cpp

```diff
@@ -17,293 +17,307 @@
     if (err == GDF_UNSUPPORTED_DTYPE || err == GDF_INVALID_API_CALL)
-        binops::jit::binary_operation(out, lhs, rhs, ope);
+      binops::jit::binary_operation(out, &temp_lhs, &temp_rhs, ope);
+
+    // TODO: Need a better way to deallocate temporary columns
+    RMM_TRY(RMM_FREE(temp_lhs.data, 0));
+    RMM_TRY(RMM_FREE(temp_rhs.data, 0));
+    if (temp_lhs.valid != nullptr) RMM_TRY(RMM_FREE(temp_lhs.valid, 0));
+    if (temp_rhs.valid != nullptr) RMM_TRY(RMM_FREE(temp_rhs.valid, 0));
+    NVCategory::destroy(
+        reinterpret_cast<NVCategory*>(temp_lhs.dtype_info.category));
+    NVCategory::destroy(
+        reinterpret_cast<NVCategory*>(temp_rhs.dtype_info.category));
+    return;
+  }
+  auto err = binops::compiled::binary_operation(out, lhs, rhs, ope);
+  if (err == GDF_UNSUPPORTED_DTYPE || err == GDF_INVALID_API_CALL)
+    binops::jit::binary_operation(out, lhs, rhs, ope);
 }
 
-void binary_operation(gdf_column* out, gdf_column* lhs, gdf_column* rhs, const std::string& ptx) {
-    // Check for null pointers in input
-    CUDF_EXPECTS((out != nullptr) && (lhs != nullptr) && (rhs != nullptr),
-        "Input pointers are null");
-
-    // Check for 0 sized data
-    if((out->size == 0) && (lhs->size == 0) && (rhs->size == 0)) return;
-    CUDF_EXPECTS((out->size == lhs->size) && (lhs->size == rhs->size),
-        "Column sizes don't match");
-
-    // Check for null data pointer
-    CUDF_EXPECTS((out->data != nullptr) &&
-                 (lhs->data != nullptr) &&
-                 (rhs->data != nullptr) , 
-        "Column data pointers are null");
-
-    // Check for datatype
-    CUDF_EXPECTS((out->dtype == lhs->dtype) && (lhs->dtype == rhs->dtype) &&
-                 (out->dtype == GDF_FLOAT32 || out->dtype == GDF_FLOAT64 || 
-                  out->dtype == GDF_INT64   || out->dtype == GDF_INT32     ),
-        "Invalid/Unsupported datatype");
-    
-    binops::binary_valid_mask_and(out->null_count, out->valid, lhs->valid, rhs->valid, rhs->size);
-
-    binops::jit::binary_operation(out, lhs, rhs, ptx);
+gdf_column binary_operation(const gdf_column& lhs, const gdf_column& rhs,
+                            const std::string& ptx, gdf_dtype output_type) {
+  CUDF_EXPECTS((lhs.size == rhs.size), "Column sizes don't match");
+
+  gdf_column output;
+  if (lhs.valid != nullptr) {
+    output = allocate_dtype(lhs, output_type);
+  } else if (rhs.valid != nullptr) {
+    output = allocate_dtype(rhs, output_type);
+  } else {
+    output = allocate_dtype(lhs, output_type);
+  }
+
+  // Check for 0 sized data
+  if ((lhs.size == 0) && (rhs.size == 0)) return output;
+  CUDF_EXPECTS((lhs.size == rhs.size), "Column sizes don't match");
+
+  // Check for null data pointer
+  CUDF_EXPECTS((lhs.data != nullptr) && (rhs.data != nullptr),
+               "Column data pointers are null");
+
+  // Check for datatype
+  CUDF_EXPECTS((lhs.dtype == GDF_FLOAT32 || lhs.dtype == GDF_FLOAT64 ||
```

Comment:
Can you break this up so it's easier to understand the different checks you are performing?